### ISSUE TITLE
Making unflatten work with Open Ownership whilst passing schema option

### ISSUE THREAD

**Issue by odscjames** [OpenDataServices/flatten-tool#225]
(This assumes #221 has been done or it won't run at all. I don't think this means #221 shouldn't be done tho.)

**Problem 1 - the easy one**

The schema for Open Ownership uses external refs, and the tool as it currently stands can't load them. Need to pass base_uri to jsonref.load like

```
--- a/flattentool/schema.py
+++ b/flattentool/schema.py
@@ -9,6 +9,9 @@ import jsonref
 from warnings import warn
 from flattentool.sheet import Sheet
 import codecs
+import pathlib
+import os
+
 
 def get_property_type_set(property_schema_dict):
     property_type = property_schema_dict.get('type', [])
@@ -92,7 +95,7 @@ class SchemaParser(object):
                 self.root_schema_dict = jsonref.loads(r.text, object_pairs_hook=OrderedDict)
             else:
                 with codecs.open(schema_filename, encoding="utf-8") as schema_file:
-                    self.root_schema_dict = jsonref.load(schema_file, object_pairs_hook=OrderedDict)
+                    self.root_schema_dict = jsonref.load(schema_file, object_pairs_hook=OrderedDict, base_uri=pathlib.Path(os.path.realpath(schema_filename)).as_uri())
         else:
             self.root_schema_dict = root_schema_dict
 ```

except I think the above is Python 3 only and we need Python 2 - WIP.

**Problem 2 - oh shit**

So the schema flag as currently designed is meant to take in a schema not for the entire file, but ONLY for the bit that is repeated in a list. eg Take a look at https://flatten-tool.readthedocs.io/en/latest/unflatten/#using-a-json-schema-with-types

The JSON Schema has no mention of "cafe", the top level list. Instead it starts at the level of one of the items in the "cafe" list.

So if I take two CSV files, which are:

```
statementID,statementType,statementDate,subject/describedByEntityStatement,interestedParty/describedByPersonStatement
fbfd0547-d0c6-4a00-b559-5c5e91c34f5c,ownershipOrControlStatement,2017-11-18,1dc0e987-5c57-4a1c-b3ad-61353b66a9b7,019a93f1-e470-42e9-957b-03559861b2e2
```

```
statementID,interests/0/type,interests/0/interestLevel,interests/0/beneficialOwnershipOrControl,interests/0/startDate,interests/0/share/exact
fbfd0547-d0c6-4a00-b559-5c5e91c34f5c,shareholding,direct,True,2016-04-06,100
```

And apply #221 and the fix in problem 1 above, and run:

    flatten-tool unflatten -f csv examples/bods-one-flatten/  -o examples/bods-one-flattened.json --no-root-list-path  --id-name=statementID --schema bods-package.json

It doesn't work - the resulting output has "exact" as as string not a number:

                "share": {
                    "exact": "100"
                }

**A Fix that isn't really a fix at all but shows the problem**

In order to get it to pick up the schema correctly and apply the typing, I have to run it with a different schema file:

    flatten-tool unflatten -f csv examples/bods-one-flatten/  -o examples/bods-one-flattened.json --no-root-list-path  --id-name=statementID --schema ownership-or-control-statement.json

Now the typing works correctly and I get:

                "share": {
                    "exact": 100
                }

However, this isn't really a fix as there are 3 types of item in the list, and here we only pass the schema for one of them - so the other 2 types will be untyped.

**Comment by odscjames** [comment]
I tried writing a new bods-package-for-unflatten.json that has just the element in the list, but has all 3 options:

```
{
  "id": "bods-package.json",
  "$schema": "http://json-schema.org/draft-04/schema#",
  "version": "0.1",
  "type": "object",
  "oneOf": [
    {
      "$ref": "entity-statement.json"
    },
    {
      "$ref": "person-statement.json"
    },
    {
      "$ref": "ownership-or-control-statement.json"
    }
  ]
}
```
But it didn't work! Not sure why.

**Comment by odscjames** [comment]
Part 1 done!

Part 2: Think I've worked out a solution in code! P.R. is currently open

**Comment by jpmckinney** [comment]
Is this closed by #235 and #237?
